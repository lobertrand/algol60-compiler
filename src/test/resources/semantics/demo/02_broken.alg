begin
    comment Array as parameter;
    integer procedure max(arr, a, b);
    value arr, a, b; string array arr; integer a, b;
    begin
        integer i, result;
        result := arr[a];
        comment For-statement;
        for i := a+1 step 1 until b do
        begin
            comment If-statement;
            if arr[i] > result then
                result := arr[i]
        end;
        max := result
    end;

    comment Array declaration and assignment;
    integer array a[1:3];
    a[0] := 6;
    a[2] := 8;
    a[4] := 3;

    comment Procedure call;
    outinteger(1, max(a, 1, 3))
end
