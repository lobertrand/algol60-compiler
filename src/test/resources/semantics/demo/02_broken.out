Algol60> Lexical analysis successful
Algol60> Syntactic analysis successful
┌───────────────────────────────────────────────┐
│ level: 0, tableNumber: 0                      │
├───────────────────────────────────────────────┤
│ Procedure: void outstring(integer, string)    │
│ Procedure: void outboolean(integer, boolean)  │
│ Procedure: void outinteger(integer, integer)  │
│ Procedure: void outreal(integer, real)        │
│ Procedure: void inreal(integer, real)         │
│ Procedure: void ininteger(integer, integer)   │
│ Procedure: integer entier(real)               │
└───────────────────────────────────────────────┘
    ┌────────────────────────────────────────────────────────────────────┐
    │ level: 1, tableNumber: 1                                           │
    ├────────────────────────────────────────────────────────────────────┤
    │ Procedure: integer max(string array, integer, integer, undefined)  │
    │ Array: integer array a[1:3]                                        │
    └────────────────────────────────────────────────────────────────────┘
        ┌─────────────────────────────────────────┐
        │ level: 2, tableNumber: 2                │
        ├─────────────────────────────────────────┤
        │ Array: string array arr[] (value param) │
        │ Variable: integer a (value param)       │
        │ Variable: integer b (value param)       │
        │ Variable: undefined extra (ref. param)  │
        │ Variable: integer max (proc. result)    │
        │ Variable: integer i                     │
        │ Variable: integer result                │
        └─────────────────────────────────────────┘
            ┌──────────────────────────┐
            │ level: 3, tableNumber: 3 │
            ├──────────────────────────┤
            └──────────────────────────┘

Algol60> Parameter mismatch at line 3: Parameter 'extra' has no type specified in specification part of procedure 'max'
     3 |     integer procedure max(arr, a, b, extra);
       |                                      ^
Algol60> Type mismatch at line 7: Cannot assign a string to 'result' of type integer
     7 |         result := arr[a];
       |         ^
Algol60> Type mismatch at line 12: Cannot perform arithmetic operation on operands of types string and integer
    12 |             if arr[i] > result then
       |                       ^
Algol60> Type mismatch at line 13: Cannot assign a string to 'result' of type integer
    13 |                 result := arr[i]
       |                 ^
Algol60> Index out of bounds at line 20: Array index '0' is out of bounds in 'a' with bounds 1:3
    20 |     a[0] := 6;
       |       ^
Algol60> Index out of bounds at line 22: Array index '4' is out of bounds in 'a' with bounds 1:3
    22 |     a[4] := 3;
       |       ^
Algol60> Parameter mismatch at line 25: Procedure 'max' expects 4 parameters but received 3
    25 |     outinteger(1, max(a, 1, 3))
       |                   ^
Algol60> Terminated with 7 semantic exceptions
